pkg/rehearse/jobs.go

```diff
@@ -222,6 +228,107 @@ func ConfigureRehearsalJobs(toBeRehearsed config.Presubmits, ciopConfigs config.
+	rehearsals := make(config.Presubmits)
+
+	for templateFile, clusterTypes := range templatesWithClusterTypes {
+		for clusterType := range clusterTypes {
+			if repo, job := pickTemplateJob(prConfigPresubmits, templateFile, clusterType); job != nil {
+				jobLogger := loggers.Job.WithFields(logrus.Fields{"target-repo": repo, "target-job": job.Name})
+
+				volumeRef, err := getVolumeIndexAndTempCMName(job.Spec.Containers[0].VolumeMounts, job.Spec.Volumes, templates)
+				if err != nil {
+					jobLogger.WithError(err).Warn(skipJobMessage)
+					continue
+				}
+
+				if volumeRef != nil {
+					job.Spec.Volumes[volumeRef.index].VolumeSource.ConfigMap.Name = volumeRef.name
+				}
+
+				rehearsals[repo] = append(rehearsals[repo], *job)
+			}
+		}
+	}
+	return rehearsals
+}
+
+type volumeRef struct {
+	index       int
+	name        string
+	templateKey string
+}
+
+func getVolumeIndexAndTempCMName(volumeMounts []v1.VolumeMount, volumes []v1.Volume, templates config.CiTemplates) (*volumeRef, error) {
+	exists, index, templateKey := hasChangedTemplateVolume(volumeMounts, volumes, templates)
+	if !exists {
+		return nil, nil
+	}
+
+	templateData, err := config.GetTemplateData(templates[templateKey])
+	if err != nil {
+		return nil, fmt.Errorf("couldn't get template's data: %s: %v", templates[templateKey], err)
+	}
+	return &volumeRef{
+		index:       index,
+		name:        config.GetTempCMName(config.GetTemplateName(templateKey), templateKey, templateData),
+		templateKey: templateKey}, nil
+}
+
+func getClusterTypeEnvIndex(envVars []v1.EnvVar) (bool, int) {
+	for index, env := range envVars {
+		if env.Name == clusterTypeEnvName {
+			return true, index
+		}
+	}
+	return false, 0
+}
+
+func GetTemplatesWithClusterTypes(templates config.CiTemplates) map[string]sets.String {
+	templatesKeys := make(map[string]sets.String)
+	for filename := range templates {
+		templatesKeys[filename] = sets.NewString([]string{"aws", "gcs", "openstack", "libvirt"}...)
+	}
+	return templatesKeys
+}
+
+func pickTemplateJob(presubmits map[string][]prowconfig.Presubmit, templateFile, clusterType string) (string, *prowconfig.Presubmit) {
+	for repo, jobs := range presubmits {
+		for _, job := range jobs {
+			if job.Agent != string(pjapi.KubernetesAgent) {
+				continue
+			}
+
+			if hasClusterType(job, clusterType) && hasTemplateFile(job, templateFile) {
+				return repo, &job
+			}
+		}
+	}
+	return "", nil
+}
+
+func hasClusterType(job prowconfig.Presubmit, clusterType string) bool {
+	for _, env := range job.Spec.Containers[0].Env {
+		if env.Name == clusterTypeEnvName && env.Value == clusterType {
+			return true
+		}
+	}
+	return false
+}
+
+func hasTemplateFile(job prowconfig.Presubmit, templateFile string) bool {
+	if job.Spec.Containers[0].VolumeMounts != nil {
+		for _, volumeMount := range job.Spec.Containers[0].VolumeMounts {
+			if volumeMount.SubPath == templateFile {
+				return true
+			}
+		}
+	}
+	return false
+}
```

Comment:
Can't these two functions be implemented using (or replaced by) `getClusterTypeEnvIndex` and `getVolumeIndexAndTempCMName`?

Reply:
I don't think so. In `hasTemplateFile()` we check a specific template file.